lua/nvim-navic/init.lua

```diff
@@ -25,6 +25,92 @@ local function request_symbol(for_buf, handler, client_id)
 	)
 end
 
+-- Return the relation of `other` to `symbol`
+--
+-- Possible values are:
+--   before
+--   around
+--   within
+--   after
+local function symbol_relation(symbol, other)
+	local s = symbol.scope
+	local o = other.scope
+
+	if
+		o["end"].line < s["start"].line
+		or (o["end"].line == s["start"].line and o["end"].character < s["start"].character)
+	then
+		return "before"
+	end
+
+	if
+		o["start"].line > s["end"].line
+		or (o["start"].line == s["end"].line and o["start"].character > s["end"].character)
+	then
+		return "after"
+	end
+
+	if
+		(
+			o["start"].line < s["start"].line
+			or (o["start"].line == s["start"].line and o["start"].character <= s["start"].character)
+		)
+		and (
+			o["end"].line > s["end"].line
+			or (o["end"].line == s["end"].line and o["end"].character >= s["end"].character)
+		)
+	then
+		return "around"
+	end
+
+	return "within"
+end
+
+-- Derive the hierarchy in a symbol list. Add all direct descendents of a
+-- symbol to a `children` property on the symbol.
+local function derive_hierarchy(symbols)
+	for _, sym in ipairs(symbols) do
```

Comment:
Looks like O(n^2) algorithm here... Can this be done faster?
Maybe sorting all the nodes by their scope and then constructing a tree could be faster?
Also, is the tree created by this identical to the tree of documentSymbols?

Reply:
> Looks like O(n^2) algorithm here... Can this be done faster?

It was already a _bit_ better than O(n^2) since the inner loop stopped after it had passed the current symbol, but I updated the loop code so that the inner loop will start just after the current symbol in the outer loop. The inner loop should now only run for the immediately applicable region around the current outer loop symbol.

> Maybe sorting all the nodes by their scope and then constructing a tree could be faster?

This function already depends on the nodes being sorted (they are, just after the initial parse). I added a doc comment to this function mentioning that assumption.

> Also, is the tree created by this identical to the tree of documentSymbols?

Hmmm...I'm not quite sure what is being asked. If it's whether this code generates the same _type_ of tree structure as is contained in documentSymbols, I believe so. Symbols contain a `children` property that points to other symbols that are direct children.

If the question is whether this code will generate an identical tree structure to the tree already contained in a set of documentSymbols, I'm not sure. The code has only been tested on SymbolInformation symbols (it's not run for servers using documentSymbols), and its correctness has only been verified observationally (by checking that navic was displaying the expected labels at various points in files). I would expect it to disagree with a documentSymbol tree occasionally, since the reason that documentSymbols explicitly describe hierarchy is because using symbol positioning to derive hierarchy isn't always reliable.

Reply:
Sorry for the long delay again...
Changes to `update_context` function wouldn't be needed if the tree constructed in this way would be exactly of same format of documentSymbol's tree. Let me try to implement something my self as well, I finally have my laptop with me again 😄

Reply:
Ah, I see what you mean. The resulting document symbol list in #55 is just the first layer of nodes in the tree, whereas it's the entire symbol list in this PR.

I actually considered the revised `update_context` worthwhile on its own because the logic is simpler, but it shouldn't have been necessary (because the symbol processing code should only have returned the first tree layer of symbols).

I any case, as long as we end up with SymbolInformation support it's a win. 🙂